### ISSUE TITLE
Import of non-Windows XP API functions

### ISSUE THREAD

**Issue by void2012** [Chuyu-Team/VC-LTL5#72]
Hi! A simple Hello World program using `<iostream>` imports WinAPI calls which aren't present on Windows XP, such as:
`LCMapStringEx`, `LCIDToLocaleName`, `InitializeSListHead`, `InitializeCriticalSectionEx`, `GetLocaleInfoEx`
VC-LTL version: `msvcrt 5.1.2600.0 (support Windows XP/2003 or leater)`
Visual Studio 2022 `version: 17.12.3`
Build tools: `LLVM (clang-cl)`
Linker minimum required option: `5.01`
C++ standard: `/std:c++latest`
Runtime library: `/MT`
Code:
`#include <iostream>
int main()
{
    std::cout << "Hello, World!\n";
}`

**Comment by mingkuang-Chuyu** [comment]
Please install [YY-Thunks](https://github.com/Chuyu-Team/YY-Thunks), or Instal VC-LTL by nuget.

**Comment by void2012** [comment]
YY-thunks seems like a product of reverse engineering. I'm not complaining, but anti-viruses aren't happy with compiled binaries.

Maybe it's possible to fork MSVC STL and remove NT6+ dependencies?

**Comment by mingkuang-Chuyu** [comment]
@void2012 Hi, you can directly delete all the `*.exe` files, they are optional.

* YY.Depends.Analyzer.exe  : Windows PE Depends Check Tool.
* MinimumRequiredVersionHelper.exe :  Fix PE Minimum Required Version（Windows 2000 support only）

**Comment by void2012** [comment]
No, the issue is not those executables. Rather, binaries compiled with VC-LTL and YY-thunks. I think there is a better approach than to reimplement NT calls with a questionable third-party reverse engineered thunk implementation.

MSVC STL was made open source in 2019 with the Apache license which allowed to use some code from llvm. Notably, from libc++.

libc++ uses three different flavors of threading system: pthreads, c11 threads and win32 threads. The latter two use NT6 WinAPI calls, while pthreads are implemented via winpthreads which has has support down to NT4.

I guess at this point we don't really care about the possible ABI change, I think one can simply reimplement MSVC STL threads with winpthreads implementation from libc++, as the license allows to do it.

An another point of interest is std::filesystem and parallel algorithms. https://github.com/gulrak/filesystem/tree/master/ implementation is fine, however symlink function is not supported prior to Windows Vista(I guess it's enough to have std::filesystem as whole but just without symlinks)

Parallel algorithms on MSVC STL work via Win32 threadpool functions, available since Windows 7. As of now, neither libc++ nor libstdc++ fully implement them but existing implementation work well with tbb library.

**Comment by MouriNaruto** [comment]
Actually, using Windows API shim instead of forking MSVC STL, will improve the source code maintenance experience and reduce the NuGet package size massively.

Kenji Mouri

**Comment by mingkuang-Chuyu** [comment]
1. We need a universal solution. And only stl compatibility with Windows 7 doesn't have much practical value, and many libraries have abandoned compatibility with Windows 7. Who's going to be Windows 7 or XP compatible?

2. We don't have the energy to maintain those libraries. Whoever does it,  but I can't.

3. Lower the barrier to entry, because ABI is fully compatible with Microsoft, and users don't have to rebuild when using VC-LTL.

In the light of the foregoing, this proposal is not considered.

**Comment by void2012** [comment]
I am considering doing my own fork of MSVC STL at this point. Can you add my fork as an **optional** dependency if I actually manage to implement it?

**Comment by mingkuang-Chuyu** [comment]
@void2012 
Your solution is similar to VC-LTL 4. You may want to consider developing based on VC-LTL 4.

https://github.com/Chuyu-Team/VC-LTL/tree/master/VC
